### PR TITLE
[WIP][Sema] Prototype: @tupleLiteral

### PR DESCRIPTION
**Pull request by Azoy**
See here: https://forums.swift.org/t/implementing-expressiblebytupleliteral-how-hard/21169

This is not done, I just wanted to get what work I've done out there before I went to sleep. *Some bugs here and there, but I'll get to those soon*. *Missing tests as well*.

Example:
```swift
struct Dog {
  let age: Int
  let name: String

  @tupleLiteral
  init(age: Int, name: String) {
    self.age = age
    self.name = name
  }
}

let x: Dog = (age: 3, name: "Sparky")

print(x) // Dog(age: 3, name: "Sparky")
```

This is currently adding a top level disjunction, so I'm not how it will affect type checker performance.

cc: @jckarter @kelvin13

### CODE REVIEW

**Comment by jckarter** [comment]
Cool. @xedin or @DougGregor would be good people to review this.

**Comment by jckarter** [comment]
@swift-ci Please test compiler performance

File: lib/Sema/CSGen.cpp [hunk 1]
```diff
@@ -1783,7 +1783,53 @@ namespace {
...
       }
 
-      return TupleType::get(elements, CS.getASTContext());
+      auto &ctx = CS.getASTContext();
+      auto locator = CS.getConstraintLocator(expr);
+      auto contextualTy = CS.getContextualType(expr);
+      auto tupleTy = TupleType::get(elements, ctx);
+
+      // Somewhat Fast path: if contextual type is a tuple, return the tuple.
+      if (contextualTy && contextualTy->is<TupleType>()) {
```
Comment (CodaFi) [inline lib/Sema/CSGen.cpp]: I would express this as a conversion, maybe even tack it onto `LiteralConformsTo`.  Going after the contextual type in CSGen is a bit too early in constraint solving.

Comment (Azoy) [reply lib/Sema/CSGen.cpp]: Does `LiteralConformsTo` make sense here? Because we aren’t introducing a new `ExpressibleBy*Literal`, so I’m not sure what the protocolTy would be.

**Comment by jckarter** [comment]
@swift-ci Please test compiler performance

**Comment by Azoy** [comment]
@jckarter this patch is unable to build the stdlib atm because there are some assumptions around the compiler that all TupleExprs return a TupleType where I return a type variable in the disjunction case.

**Comment by jckarter** [comment]
In the case when the expression resolves to an initializer, you probably want to rewrite the AST as a form of `ApplyExpr` invoking the initializer.

File: lib/Sema/CSGen.cpp (hunk 1 above)
Comment (CodaFi) [reply lib/Sema/CSGen.cpp]: Sorry, that was unclear.  Let me rephrase: This constraint makes the most sense as a conversion constraint in the style of `LiteralConformsTo` for an infinite family of ad-hoc `ExpressibleByTupleOfXXXLiteral` protocols.  You could emit such a constraint and basically nuke everything down below here.  Then your corresponding `CSSimplify` rule breaks this down into an argument match against constructors.

**Comment by jckarter** [comment]
It looks like there are still issues with CI beyond anything specific to this patch, alas.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?
